src/core/functions/internal_functions/system/MultiSelectModal.ts

```diff
@@ -0,0 +1,229 @@
+    private chipsContainer: HTMLDivElement;
+    private textInput: TextComponent;
+
+    constructor(
+        private text_items: string[] | ((item: T) => string),
+        private items: T[],
+        private unrestricted: boolean,
+        title: string,
+        private limit?: number
+    ) {
+        super(app);
+        this.initializeModal(items, title);
+    }
+
+    initializeModal(items: T[], title: string) {
+        // If item isn't of type string or items array is empty, then unrestricted is false
+        if (items.length === 0 || typeof items[0] !== "string") {
+            this.unrestricted = false;
+        }
+
+        // Map each item to its text representation
+        items.forEach((item) =>
+            this.mappedItems.set(this.getItemText(item), item)
+        );
+
+        // Create Elements
+        this.titleEl.setText(title);
+        this.chipsContainer = this.contentEl.createDiv();
+        this.textInput = new TextComponent(this.contentEl);
+        this.suggestionsContainer = this.contentEl.createDiv();
+
+        // Add styles
+        this.modalEl.addClass("templater-multiselect-modal");
+        this.contentEl.addClass("templater-multiselect-container");
+        this.chipsContainer.addClass("templater-multiselect-chips");
+        this.suggestionsContainer.addClass("templater-multiselect-suggestions");
+    }
+
+    onOpen(): void {
+        this.textInput.inputEl.addEventListener(
+            "keydown",
+            (evt: KeyboardEvent) => {
+                switch (evt.key) {
+                    case "Tab":
+                        this.addItemFromSuggestionsOrInputField();
+                        break;
+
+                    case "Backspace":
+                        if (!this.textInput.getValue()) this.removeLastItem();
+                        break;
+
+                    case "ArrowDown":
+                        this.navigateSelection("DOWN");
+                        break;
+
+                    case "ArrowUp":
+                        this.navigateSelection("UP");
+                        break;
+
+                    case "Enter":
+                        this.submitted = true;
+                        this.close();
+                        break;
+                }
+            }
+        );
+
+        this.textInput.onChange(() => this.renderSuggestions());
+        this.renderSuggestions();
+    }
+
+    addItemFromSuggestionsOrInputField() {
+        const inputText = this.textInput.getValue().trim();
+        const selected = this.suggestionsContainer.querySelector(
+            this.SELECTOR_SELECTED
+        );
+        if (selected) this.addItem(selected.textContent!);
+        else if (inputText && this.unrestricted) this.addItem(inputText);
+    }
+
+    addItem(value: string) {
+        this.selectedItems.add(value);
+        this.addChip(value);
+        this.textInput.setValue("");
+        this.renderSuggestions();
+    }
+
+    removeLastItem() {
```

Comment:
When this function is called, it isn't adding back the item to the list of available suggestions. I think it should be added back as an available item to select.

Reply:
Do you mean `removeLastItem()` function? 
It removes the selected item and re-renders suggestions by calling `renderSuggestions()`. 
`renderSuggestions()` function filters out items which are selected and renders suggestions which are not selected. 
Since we remove the selected item in `removeLastItem()` it should be available for suggestions.

If you mean something else, can you please provide me more details.